numba/analysis.py

```diff
@@ -259,3 +260,179 @@ def find_top_level_loops(cfg):
     for loop in cfg.loops().values():
         if loop.header not in blocks_in_loop:
             yield loop
+
+# Functions to manipulate IR
+def dead_branch_prune(func_ir, called_args):
+    """
+    Removes dead branches based on constant inference from function args.
+    This directly mutates the IR.
+
+    func_ir is the IR
+    called_args are the actual arguments with which the function is called
+    """
+
+    DEBUG = 0
+
+    def find_branches(func_ir):
+        # find *all* branches
+        branches = []
+        for idx, blk in func_ir.blocks.items():
+            tmp = [_ for _ in blk.find_insts(cls=ir.Branch)]
+            store = dict()
+            for branch in tmp:
+                store['branch'] = branch
+                expr = blk.find_variable_assignment(branch.cond.name)
+                if expr is not None:
+                    val = expr.value
+                    if isinstance(val, ir.Expr) and val.op == 'binop':
+                        store['op'] = val
+                        args = [val.lhs, val.rhs]
+                        store['args'] = args
+                        store['block'] = blk
+                        branches.append(store)
+        return branches
+
+    def prune(func_ir, branch, all_branches, *conds):
+        # this prunes a given branch and fixes up the IR
+        lhs_cond, rhs_cond = conds
+        take_truebr = branch['op'].fn(lhs_cond, rhs_cond)
+        cond = branch['branch']
+        if take_truebr:
+            keep = cond.truebr
+            kill = cond.falsebr
+        else:
+            keep = cond.falsebr
+            kill = cond.truebr
+
+        cfg = compute_cfg_from_blocks(func_ir.blocks)
+        if DEBUG > 0:
+            print("Pruning %s" % kill, branch['branch'], lhs_cond, rhs_cond, branch['op'].fn)
+            if DEBUG > 1:
+                from pprint import pprint
+                cfg.dump()
+
+        # only prune branches to blocks that have a single access route, this is
+        # conservative
+        kill_count = 0
+        for targets in cfg._succs.values():
+            if kill in targets:
+                kill_count += 1
+                if kill_count > 1:
+                    if DEBUG > 0:
+                        print("prune rejected")
+                    return
+
+        # remove dominators that are not on the backbone
+        dom = cfg.dominators()
+        postdom = cfg.post_dominators()
+        backbone = cfg.backbone()
+        rem = []
+
+        for idx, doms in dom.items():
```

Comment:
Remove the block if it is dominating _some_ block and is not in the backbone? 
I think a block can be removed if it doesn't have any predecessors left. Backbone check is just an assertion showing something has gone wrong. 

Also, removing blocks can be done once after the big loop when all constant branches are resolved.

Reply:
Yes, thanks, done. Turns out that just switching branches for jumps and then computing the dead nodes in the CFG is all that's needed.